src/documentation/0000-socket-programming-with-net-module/index.md

```diff
@@ -0,0 +1,136 @@
+
+Node.js exposes extensive API's for implementing these protocols so developers can create socket programming applications. The most common example is the `http` module. [HTTP](https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol) exists on the Application layer of the OSI Model; if you want to learn more about the `http` module, read the Node.js [http documentation](https://nodejs.org/api/http.html) or the Nodejs.dev [guide](https://nodejs.dev/the-nodejs-http-module). Another socket programming, Node.js module is the `dgram` module for [UDP](https://en.wikipedia.org/wiki/User_Datagram_Protocol) servers. UDP exists on the Transport layer; for more information about this technology read the Node.js [dgram documentaion](https://nodejs.org/api/dgram.html). More importantly, the `net` module, which this guide is all about, provides an extensive API for a stream-based TCP server. [TCP](https://en.wikipedia.org/wiki/Transmission_Control_Protocol) exists on the Transport layer of the OSI Model and is what HTTP is implemented on.
+
+> TCP is a reliable, ordered, and error-checked delivery of a stream of bytes between applications communicating via an internet protocol (IP) network.
+>
+> ~ Wikipedia
+
+## Net Module - Getting Started
+
+The key part of the `net` module is the `net.Socket` class. A socket instance in Node.js is both a duplex stream and an event emitter. This means it is readable, writable, and can utilize the event loop. Sockets can be created by the user and by Node.js, and they are the programmatic object for communicating. Sockets must connect to a `net.Server` class instance; a socket connection to a server is often referred to as a _client connection_. Every socket connection is considered "private" in the sense that the only entities with access to the connection is the server and the client. A server instance **can** have multiple client connections, but will maintain separate communication streams with each connection.
+
+Lets look at the simplest example of a TCP server, an echo server. An _echo server_ is one which returns the same message sent to it.
+
+```js
+const net = require('net')
+
+// Create a TCP server using the `net.createServer()` method
+// The callback method is an event listener for the 'connection' event
+// This method is passed a single parameter, the client socket connection
+const server = net.createServer(client => {
+
+  console.log('Client connected!')
+
+  // The client is a stream, set the data encoding to utf8
+  client.setEncoding('utf8')
+
+  // Manual implementation of an echo server:
+  client.on('data', data => { // data is a UTF-8 string
+    console.log(data) // log the message from the client to the server
+    client.write(data) // write the same message back to the client
+  })
+
+  // Automatic implementation of an echo server:
+  client.pipe(client) // client is a stream, use the Stream.pipe prototype method
```

Comment:
```suggestion
  // client is a stream, use the Stream.pipe prototype method to connect it's
  // read stream to it's write stream
  // client.pipe(client)
```
I think we should leave this commented out as the current code if copy pasted is going to double up the messages.